packages/router/src/router.tsx

```diff
@@ -1,41 +1,221 @@
+/**
+ * External dependencies
+ */
+import RouteRecognizer from 'route-recognizer';
+import { createBrowserHistory } from 'history';
+
 /**
  * WordPress dependencies
  */
 import {
 	createContext,
 	useContext,
 	useSyncExternalStore,
+	useMemo,
 } from '@wordpress/element';
+import {
+	addQueryArgs,
+	getQueryArgs,
+	getPath,
+	buildQueryString,
+} from '@wordpress/url';
 
 /**
  * Internal dependencies
  */
-import history from './history';
-import type { EnhancedHistory } from './history';
+import type { ReactNode } from 'react';
+
+const history = createBrowserHistory();
+interface Route {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+}
 
-const RoutesContext = createContext< Location | null >( null );
-const HistoryContext = createContext< EnhancedHistory >( history );
+type LocationWithQuery = Location & {
+	query?: Record< string, any >;
+};
+
+interface Match {
+	name: string;
+	path: string;
+	areas: Record< string, ReactNode >;
+	widths: Record< string, number >;
+	query?: Record< string, any >;
+	params?: Record< string, any >;
+}
+
+export type BeforeNavigate = ( arg: {
+	path: string;
+	query: Record< string, any >;
+} ) => {
+	path: string;
+	query: Record< string, any >;
+};
+
+interface Config {
+	pathArg: string;
+	beforeNavigate?: BeforeNavigate;
+}
+
+export interface NavigationOptions {
+	transition?: string;
+	state?: Record< string, any >;
+}
+
+const RoutesContext = createContext< Match | null >( null );
+export const ConfigContext = createContext< Config >( { pathArg: 'p' } );
+
+const locationMemo = new WeakMap();
+function getLocationWithQuery() {
+	const location = history.location;
+	let locationWithQuery = locationMemo.get( location );
+	if ( ! locationWithQuery ) {
+		locationWithQuery = {
+			...location,
+			query: Object.fromEntries( new URLSearchParams( location.search ) ),
+		};
+		locationMemo.set( location, locationWithQuery );
+	}
+	return locationWithQuery;
+}
 
 export function useLocation() {
```

Comment:
Should we introduce a check and bail if the context doesn't exist?

Reply:
Could be a `warning()` or could be an actual error thrown - I lean towards the warning as it's an error targeted towards developers and just throwing might not be the best experience, given how we catch those errors in the editor.